core/actor/src/main/scala/net/liftweb/actor/LAFuture.scala

```diff
@@ -376,40 +441,21 @@ object LAFuture {
    * returned future with an Empty
    */
   def collectAll[T](future: LAFuture[Box[T]]*): LAFuture[Box[List[T]]] = {
-    val result = new LAFuture[Box[List[T]]]
-    if (future.isEmpty) {
-      result.satisfy(Full(Nil))
-    } else {
-      val sync = new Object
-      val len = future.length
-      val vals = new collection.mutable.ArrayBuffer[Box[T]](len)
-      // pad array so inserts at random places are possible
-      for (i <- 0 to len) { vals.insert(i, Empty) }
-      var gotCnt = 0
-
-      future.toList.zipWithIndex.foreach {
-        case (f, idx) =>
-          f.foreach {
-            vb => sync.synchronized {
-              vb match {
-                case Full(v) => {
-                  vals.insert(idx, Full(v))
-                  gotCnt += 1
-                  if (gotCnt >= len) {
-                    result.satisfy(Full(vals.toList.flatten))
-                  }
-                }
-
-                case eb: EmptyBox => {
-                  result.satisfy(eb)
-                }
-              }
-            }
-          }
-      }
-    }
-
-    result
+    collect[Box[T], Box[List[T]]](
+      { (value, result, values, index) =>
+        value match {
+          case Full(realValue) =>
+            values.insert(index, Full(Full(realValue)))
+          case other: EmptyBox =>
+            result.satisfy(other)
+        }
+      },
+      { (valueBox, result, values, index) => result.fail(valueBox) },
+      { (result: LAFuture[Box[List[T]]], values: ArrayBuffer[Box[Box[T]]]) =>
+        result.satisfy(Full(values.toList.flatten.flatten))
+      },
```

Comment:
Ditto here on somehow indicating how each of these functions are mapped to the underlying collect implementation without having to refer back to the positional order of things in that function.

Though here I feel more strongly that these functions should be their own `def`s nested inside `collectAll`